File: content/snippets/scala3/interview_lcci/02/07_intersection_of_two_linked_lists.scala
object Lcci0207IntersectionOfTwoLinkedLists:
  final case class ListNode(var value: Int, var next: ListNode | Null = null)

  def buildList(nums: Vector[Int]): ListNode | Null =
    val dummy = ListNode(0, null)
    var tail: ListNode = dummy
    nums.foreach { n =>
      tail.next = ListNode(n, null)
      tail = tail.next.asInstanceOf[ListNode]
    }
    dummy.next

  def getIntersectionNode(a: ListNode | Null, b: ListNode | Null): ListNode | Null =
    if (a == null || b == null) return null

    var p1 = a
    var p2 = b
    while (p1 != p2) {
      p1 = if (p1 != null) p1.asInstanceOf[ListNode].next else b
      p2 = if (p2 != null) p2.asInstanceOf[ListNode].next else a
    }
    p1

  @main def main(): Unit =
    val common = buildList(Vector(8, 10))
    val a = buildList(Vector(3, 1, 5, 9))
    val b = buildList(Vector(4, 6))

    var tailA = a
    while (tailA != null && tailA.asInstanceOf[ListNode].next != null) {
      tailA = tailA.asInstanceOf[ListNode].next
    }
    if (tailA != null) {
      tailA.asInstanceOf[ListNode].next = common
    }

    var tailB = b
    while (tailB != null && tailB.asInstanceOf[ListNode].next != null) {
      tailB = tailB.asInstanceOf[ListNode].next
    }
    if (tailB != null) {
      tailB.asInstanceOf[ListNode].next = common
    }

    val node = getIntersectionNode(a, b)
    println(if (node != null) node.asInstanceOf[ListNode].value else "null")

File: content/snippets/cpp/interview_lcci/02/07_intersection_of_two_linked_lists.cpp
#include <iostream>
#include <vector>

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int v) : val(v), next(nullptr) {}
};

ListNode* buildList(const std::vector<int>& nums) {
    ListNode dummy(0);
    ListNode* tail = &dummy;
    for (int n : nums) {
        tail->next = new ListNode(n);
        tail = tail->next;
    }
    return dummy.next;
}

ListNode* getIntersectionNode(ListNode* a, ListNode* b) {
    if (!a || !b) return nullptr;
    ListNode* p1 = a;
    ListNode* p2 = b;
    while (p1 != p2) {
        p1 = p1 ? p1->next : b;
        p2 = p2 ? p2->next : a;
    }
    return p1;
}

int main() {
    ListNode* common = buildList({8, 10});

    ListNode* a = buildList({3, 1, 5, 9});
    ListNode* b = buildList({4, 6});

    ListNode* tailA = a;
    while (tailA->next) tailA = tailA->next;
    tailA->next = common;

    ListNode* tailB = b;
    while (tailB->next) tailB = tailB->next;
    tailB->next = common;

    ListNode* intersection = getIntersectionNode(a, b);
    if (intersection) {
        std::cout << intersection->val << '\n';
    } else {
        std::cout << "null\n";
    }
    return 0;
}
